builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,577 @@
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job.")
+		return nil
+	} else if deploy.JobVersion != *job.Version {
+		st.Update("Job version does not match deployment's job version.")
+		return nil
+	} else if deploy.Status != "running" {
+		st.Update("Deployment for job is no longer running.")
+		return nil
+	}
+
+	canaryDeployment := false
+	groupsToPromote := make([]string, len(deploy.TaskGroups))
+	for taskGroupName, taskGroup := range deploy.TaskGroups {
+		if r.config.Groups != nil {
+			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries > 0 {
+				canaryDeployment = true
+				groupsToPromote = append(groupsToPromote, taskGroupName)
+				continue
+			}
+		} else if taskGroup.DesiredCanaries > 0 {
+			canaryDeployment = true
+			// If no groups to promote are specified in the config, promote all groups
+			//   that have canaries
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
+	}
+	if !canaryDeployment {
+		log.Info("Canaries not detected")
+		return nil
+	}
+
+	var currentTaskGroupState *api.DeploymentState
+	for _, group := range groupsToPromote {
+		if group != "" {
+			st.Update(fmt.Sprintf("Checking task group: %s", group))
+			// TODO: Update to pair a task group's name with its healthy deadline, so
+			//       we can set the deadline accordingly
+			//       d := time.Now().Add(time.Nanosecond * time.Duration(*job.TaskGroups[indexOfTaskGroupInSliceOfTaskGroupsOfJob].Update.HealthyDeadline))
+			d := time.Now().Add(time.Minute * time.Duration(5))
+			log.Debug(fmt.Sprintf("Healthy deadline: %s", d.String()))
+			ctx, cancel := context.WithDeadline(ctx, d)
+			defer cancel()
+			ticker := time.NewTicker(5 * time.Second)
+			groupHealthy := false
+			for !groupHealthy {
+				currentTaskGroupState = deploy.TaskGroups[group]
+				if currentTaskGroupState.HealthyAllocs < len(currentTaskGroupState.PlacedCanaries) {
+					select {
+					case <-ticker.C:
+					case <-ctx.Done(): // cancelled
+						return status.Errorf(codes.Aborted, "Healthy deadline reached: %s", ctx.Err())
+					}
+					deploy, _, err = jobClient.LatestDeployment(*job.ID, q)
+					currentTaskGroupState = deploy.TaskGroups[group]
+					log.Info(fmt.Sprintf("Task group not healthy: %s", group))
+				} else {
+					groupHealthy = true
+				}
+			}
+		}
+	}
+
+	wq := &api.WriteOptions{
+		Namespace: *job.Namespace,
+	}
+
+	var u *api.DeploymentUpdateResponse
+	if r.config.FailDeployment {
+		u, _, err = deploymentClient.Fail(deploy.ID, wq)
+	} else {
+		u, _, err = deploymentClient.PromoteGroups(deploy.ID, groupsToPromote, wq)
+	}
+	if err != nil {
+		return err
+	}
+
+	st.Update(fmt.Sprintf("Monitoring evaluation %s", string(u.EvalID)))
```

Comment:
StepGroup I believe supports string formatting through its parameters, so you shouldn't need `fmt` here, should be able to do `"Monitoring evaluation %q", string(u.EvalID)` without the `fmt.Sprintf`!

Reply:
I might of missed this else where, so if there are other places, it would be good to get those updated too!

Reply:
Thanks @briancain! Since `st` here is of type `terminal.Status`, did you mean to convert this (and other usages of Status.Update) to use `StepGroup`?

Also just for clarification, I used `terminal.Status` here instead of `StepGroup` as I saw other comments mentioning to not use `StepGroup` until #1536 is resolved.

Reply:
Yeah, so this is a bit confusing. This `st.Update` is a StepGroup, and `Status` is apart of that implementation I believe. The issue linked basically says we should avoid setting `Status` after a `StepGroup` output due to that bug, but I don't think we're doing that here! Just some extra info for you right now as to what that issue is about 😄 

But what I was suggesting is this should be possible, no need to use `fmt.Sprintf`:

```golang
st.Update("Monitoring evaluation %s", string(u.EvalID))
```

Reply:
@briancain I actually gave that a try, but it wouldn't compile 🤔 resulting in this warning:
```
too many arguments in call to st.Update
	have (string, string)
	want (string)
```

Dug through the core code, and _think_ the root cause is that `runnerUIStatus` (which is what I'm using in `resourceJobCreate`) does not do the formatting for us already...

```go
func (u *runnerUIStatus) Step(status string, msg string) {
	u.b.mu.Lock()
	defer u.b.mu.Unlock()

	if u.b.evc == nil {
		return
	}
```

https://github.com/hashicorp/waypoint/blob/main/internal/runner/ui.go#L255-L262

...but `runnerUISGStep` does!

```go
func (u *runnerUISGStep) Update(str string, args ...interface{}) {
	msg := fmt.Sprintf(str, args...)

	u.sg.ui.mu.Lock()
	defer u.sg.ui.mu.Unlock()
```

https://github.com/hashicorp/waypoint/blob/main/internal/runner/ui.go#L416-L421

I saw examples in other plugins which use basic concatenation with `status.Update` instead of string formatting, could do that here as well! Let me know what you think. 😄 

```go
	st.Update("Deleting Lambda function version " + deployment.Version)
```

Reply:
Oops - ok that makes sense. I guess it was originally a `Status` and not a `StepGroup`! I think your example below with the string concatenation makes sense 👍🏻 Sorry for the mixup!